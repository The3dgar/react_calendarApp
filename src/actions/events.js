import Swal from "sweetalert2";
import { fetchWithToken } from "../helper/fetch";
import { prepareEvents } from "../helper/prepareEvents";
import { types } from "../types/types";

// disparar si se graba en la base de datos
export const eventStartAddNew = (event) => {
  return async (dispatch, getState) => {
    const { uid, name } = getState().auth;
    try {
      const resp = await fetchWithToken("/events", event, "POST");
      const body = await resp.json();

      if (body.ok) {
        event.id = body.event.id;
        event.user = {
          _id: uid,
          name,
        };
        dispatch(eventAddNew(event));
      }
    } catch (error) {
      console.log(error);
    }
  };
};
const eventAddNew = (event) => ({
  type: types.eventAddNew,
  payload: event,
});

export const eventSetActive = (event) => ({
  type: types.eventSetActive,
  payload: event,
});

export const eventClearActiveEvent = () => ({
  type: types.eventClearActiveEvent,
});

export const eventStartLoading = () => {
  return async (dispatch) => {
    try {
      const resp = await fetchWithToken("events");
      const body = await resp.json();
      console.log(body);
      const events = prepareEvents(body.events);
      dispatch(eventLoaded(events));
    } catch (error) {
      console.log(error);
    }
  };
};

const eventLoaded = (events) => ({
  type: types.eventLoaded,
  payload: events,
});

export const eventStartUpdate = (event) => {
  return async (dispatch) => {
    try {
      const resp = await fetchWithToken(`/events/${event.id}`, event, "PUT");
      const body = await resp.json();
      if (body.ok) {
        dispatch(eventUpdated(event));
      } else {
        Swal.fire("Error", body.msg, "error");
      }
    } catch (error) {
      console.log(error);
    }
  };
};

const eventUpdated = (event) => ({
  type: types.eventUpdated,
  payload: event,
});

export const eventStartDelete = (event) => {
  return async (dispatch, getState) => {
    const { activeEvent: event } = getState().calendar;
    try {
      const resp = await fetchWithToken(`/events/${event.id}`, event, "DELETE");
      const body = await resp.json();
      if (body.ok) {
        dispatch(eventDeleted(event));
      } else {
        Swal.fire("Error", body.msg, "error");
      }
    } catch (error) {
      console.log(error);
    }
  };
};

const eventDeleted = () => ({
  type: types.eventDeleted,
});

export const eventLogout = () => ({
  type: types.eventLogout
})
