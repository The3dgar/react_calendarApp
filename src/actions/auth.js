import Swal from "sweetalert2";
import { fetchWithOutToken, fetchWithToken } from "../helper/fetch";
import { types } from "../types/types";
import { eventLogout } from "./events";

export const startLogin = (email, password) => {
  // dispatch gracias al thunk
  return async (dispatch) => {
    const resp = await fetchWithOutToken("auth", { email, password }, "POST");
    const body = await resp.json();

    if (body.ok) {
      const { token, uid, name } = body;
      localStorage.setItem("token", token);
      localStorage.setItem("token_init", new Date().getTime());

      dispatch(login({ uid, name }));
    } else {
      Swal.fire("Error", body.msg, "error");
    }
  };
};

const login = (user) => ({
  type: types.authLogin,
  payload: user,
});

export const startRegister = (email, password, name) => {
  return async (dispatch) => {
    const resp = await fetchWithOutToken(
      "auth/new",
      { email, password, name },
      "POST"
    );
    const body = await resp.json();

    if (body.ok) {
      const { token, uid, name } = body;
      localStorage.setItem("token", token);
      localStorage.setItem("token_init", new Date().getTime());

      dispatch(login({ uid, name }));
    } else {
      Swal.fire("Error", body.msg, "error");
    }
  };
};

export const startLogout = () => {
  return dispatch => {
    localStorage.clear()
    dispatch(eventLogout())
    dispatch(logout())
  }
}

const logout = () => ({type: types.authLogout})

export const startChecking  = ()  => {
  return async dispatch => {
    const token = localStorage.getItem("token")

    if(!token) return dispatch(checkingFinish())
    
    const resp = await fetchWithToken("auth/renew");
    const body = await resp.json();
    if (body.ok) {
      const { token, uid, name } = body;
      localStorage.setItem("token", token);
      localStorage.setItem("token_init", new Date().getTime());

      dispatch(login({ uid, name }));
    } else {
      Swal.fire("Errores", body.msg, "error");
      dispatch(checkingFinish())
    }
  }
}

const checkingFinish = () => ({
  type: types.authCheckingFinish
})