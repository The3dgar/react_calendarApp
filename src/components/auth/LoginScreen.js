import React from "react";
import "./login.css";
import { useForm } from "../../hooks/useForm";
import { useDispatch } from "react-redux";
import { startLogin, startRegister } from "../../actions/auth";
import Swal from "sweetalert2";
export const LoginScreen = () => {
  const [formLoginValues, handleLoginInputChange] = useForm({
    lEmail: "edgarolivar16@gmail.cl",
    lPassword: "123456",
  });
  const [formRegisterValues, handleRegisterInputChange] = useForm({
    rName: "Moises",
    rEmail: "moisesolivar16@gmail.cl",
    rPassword: "123456",
    rPassword2: "123456",
  });

  const { lEmail, lPassword } = formLoginValues;
  const { rName, rEmail, rPassword, rPassword2 } = formRegisterValues;

  const dispatch = useDispatch();
  const handleLogin = (e) => {
    e.preventDefault();
    dispatch(startLogin(lEmail, lPassword));
  };

  const handleRegister = e => {
    e.preventDefault();

    if(rPassword !== rPassword2) return Swal.fire("Warning", "Passwords dont match", "warning")
    dispatch(startRegister(rEmail, rPassword, rName))

  }

  return (
    <div className="container login-container">
      <div className="row">
        <div className="col-md-6 login-form-1">
          <h3>Login</h3>
          <form onSubmit={handleLogin}>
            <div className="form-group">
              <input
                type="text"
                className="form-control"
                placeholder="Correo"
                name="lEmail"
                value={lEmail}
                onChange={handleLoginInputChange}
              />
            </div>
            <div className="form-group">
              <input
                type="password"
                className="form-control"
                placeholder="Contraseña"
                name="lPassword"
                value={lPassword}
                onChange={handleLoginInputChange}
              />
            </div>
            <div className="form-group">
              <input type="submit" className="btnSubmit" value="Login" />
            </div>
          </form>
        </div>

        <div className="col-md-6 login-form-2">
          <h3>Register</h3>
          <form onSubmit={handleRegister}>
            <div className="form-group">
              <input
                type="text"
                className="form-control"
                placeholder="Nombre"
                name="rName"
                value={rName}
                onChange={handleRegisterInputChange}
              />
            </div>
            <div className="form-group">
              <input
                type="email"
                className="form-control"
                placeholder="Correo"
                name="rEmail"
                value={rEmail}
                onChange={handleRegisterInputChange}
              />
            </div>
            <div className="form-group">
              <input
                type="password"
                className="form-control"
                placeholder="Contraseña"
                name="rPassword"
                value={rPassword}
                onChange={handleRegisterInputChange}
              />
            </div>

            <div className="form-group">
              <input
                type="password"
                className="form-control"
                placeholder="Repita la contraseña"
                name="rPassword2"
                value={rPassword2}
                onChange={handleRegisterInputChange}
              />
            </div>

            <div className="form-group">
              <input type="submit" className="btnSubmit" value="Register" />
            </div>
          </form>
        </div>
      </div>
    </div>
  );
};
