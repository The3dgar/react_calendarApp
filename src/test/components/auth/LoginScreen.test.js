import React from "react";
import { mount } from "enzyme";
import { Provider } from "react-redux";
import configureStore from "redux-mock-store";
import thunk from "redux-thunk";
import "@testing-library/jest-dom";

import { LoginScreen } from "../../../components/auth/LoginScreen";
import { startLogin, startRegister } from "../../../actions/auth";
import Swal from "sweetalert2";

jest.mock("../../../actions/auth", () => ({
  startLogin: jest.fn(),
  startRegister: jest.fn()
}));

jest.mock("sweetalert2", () => ({
  fire: jest.fn(),
}));

const middlewares = [thunk];
const mockStore = configureStore(middlewares);

const initState = {};
const store = mockStore(initState);
store.dispatch = jest.fn();

const wrapper = mount(
  <Provider store={store}>
    <LoginScreen />
  </Provider>
);

describe("testing in <LoginScreen/>", () => {

  beforeEach(()=>{
    jest.clearAllMocks()
  })
  test("should snap correctly", () => {
    expect(wrapper).toMatchSnapshot();
  });

  test("should dispatch login", () => {
    wrapper.find("form").at(0).prop("onSubmit")({
      preventDefault() {},
    });

    expect(startLogin).toHaveBeenCalledWith("edgarolivar16@gmail.cl", "123456");
  });

  test("should dispatch error if password1 != password2", async () => {
    // start register dont be called

    // swal.fire called with "Warning", "Passwords dont match", "warning"

    wrapper.find('input[name="rPassword"]').simulate("change", {
      target: {
        name: "rPassword",
        value: "p1",
      },
    });
    wrapper.find('input[name="rPassword2"]').simulate("change", {
      target: {
        name: "rPassword2",
        value: "p2",
      },
    });
    expect(wrapper).toMatchSnapshot();
    wrapper.find("form").at(1).prop("onSubmit")({ preventDefault() {} });

    expect(startRegister).toBeCalledTimes(0);
    expect(Swal.fire).toHaveBeenCalledWith(
      "Warning",
      "Passwords dont match",
      "warning"
    );
  });

  test('should dispatch register with the same password', () => {
    wrapper.find('input[name="rPassword"]').simulate("change", {
      target: {
        name: "rPassword",
        value: "1234567",
      },
    });
    wrapper.find('input[name="rPassword2"]').simulate("change", {
      target: {
        name: "rPassword2",
        value: "1234567",
      },
    });
    expect(wrapper).toMatchSnapshot();
    wrapper.find("form").at(1).prop("onSubmit")({ preventDefault() {} });

    expect(startRegister).toBeCalled();
    expect(Swal.fire).not.toHaveBeenCalled()
  })
  
});
