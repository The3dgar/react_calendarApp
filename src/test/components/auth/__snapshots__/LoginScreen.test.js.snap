// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testing in <LoginScreen/> should dispatch error if password1 != password2 1`] = `
<div
  className="container login-container"
>
  <div
    className="row"
  >
    <div
      className="col-md-6 login-form-1"
    >
      <h3>
        Login
      </h3>
      <form
        onSubmit={[Function]}
      >
        <div
          className="form-group"
        >
          <input
            className="form-control"
            name="lEmail"
            onChange={[Function]}
            placeholder="Correo"
            type="text"
            value="edgarolivar16@gmail.cl"
          />
        </div>
        <div
          className="form-group"
        >
          <input
            className="form-control"
            name="lPassword"
            onChange={[Function]}
            placeholder="Contraseña"
            type="password"
            value="123456"
          />
        </div>
        <div
          className="form-group"
        >
          <input
            className="btnSubmit"
            type="submit"
            value="Login"
          />
        </div>
      </form>
    </div>
    <div
      className="col-md-6 login-form-2"
    >
      <h3>
        Register
      </h3>
      <form
        onSubmit={[Function]}
      >
        <div
          className="form-group"
        >
          <input
            className="form-control"
            name="rName"
            onChange={[Function]}
            placeholder="Nombre"
            type="text"
            value="Moises"
          />
        </div>
        <div
          className="form-group"
        >
          <input
            className="form-control"
            name="rEmail"
            onChange={[Function]}
            placeholder="Correo"
            type="email"
            value="moisesolivar16@gmail.cl"
          />
        </div>
        <div
          className="form-group"
        >
          <input
            className="form-control"
            name="rPassword"
            onChange={[Function]}
            placeholder="Contraseña"
            type="password"
            value="p1"
          />
        </div>
        <div
          className="form-group"
        >
          <input
            className="form-control"
            name="rPassword2"
            onChange={[Function]}
            placeholder="Repita la contraseña"
            type="password"
            value="p2"
          />
        </div>
        <div
          className="form-group"
        >
          <input
            className="btnSubmit"
            type="submit"
            value="Register"
          />
        </div>
      </form>
    </div>
  </div>
</div>
`;

exports[`testing in <LoginScreen/> should dispatch register with the same password 1`] = `
<div
  className="container login-container"
>
  <div
    className="row"
  >
    <div
      className="col-md-6 login-form-1"
    >
      <h3>
        Login
      </h3>
      <form
        onSubmit={[Function]}
      >
        <div
          className="form-group"
        >
          <input
            className="form-control"
            name="lEmail"
            onChange={[Function]}
            placeholder="Correo"
            type="text"
            value="edgarolivar16@gmail.cl"
          />
        </div>
        <div
          className="form-group"
        >
          <input
            className="form-control"
            name="lPassword"
            onChange={[Function]}
            placeholder="Contraseña"
            type="password"
            value="123456"
          />
        </div>
        <div
          className="form-group"
        >
          <input
            className="btnSubmit"
            type="submit"
            value="Login"
          />
        </div>
      </form>
    </div>
    <div
      className="col-md-6 login-form-2"
    >
      <h3>
        Register
      </h3>
      <form
        onSubmit={[Function]}
      >
        <div
          className="form-group"
        >
          <input
            className="form-control"
            name="rName"
            onChange={[Function]}
            placeholder="Nombre"
            type="text"
            value="Moises"
          />
        </div>
        <div
          className="form-group"
        >
          <input
            className="form-control"
            name="rEmail"
            onChange={[Function]}
            placeholder="Correo"
            type="email"
            value="moisesolivar16@gmail.cl"
          />
        </div>
        <div
          className="form-group"
        >
          <input
            className="form-control"
            name="rPassword"
            onChange={[Function]}
            placeholder="Contraseña"
            type="password"
            value="1234567"
          />
        </div>
        <div
          className="form-group"
        >
          <input
            className="form-control"
            name="rPassword2"
            onChange={[Function]}
            placeholder="Repita la contraseña"
            type="password"
            value="1234567"
          />
        </div>
        <div
          className="form-group"
        >
          <input
            className="btnSubmit"
            type="submit"
            value="Register"
          />
        </div>
      </form>
    </div>
  </div>
</div>
`;

exports[`testing in <LoginScreen/> should snap correctly 1`] = `
<div
  className="container login-container"
>
  <div
    className="row"
  >
    <div
      className="col-md-6 login-form-1"
    >
      <h3>
        Login
      </h3>
      <form
        onSubmit={[Function]}
      >
        <div
          className="form-group"
        >
          <input
            className="form-control"
            name="lEmail"
            onChange={[Function]}
            placeholder="Correo"
            type="text"
            value="edgarolivar16@gmail.cl"
          />
        </div>
        <div
          className="form-group"
        >
          <input
            className="form-control"
            name="lPassword"
            onChange={[Function]}
            placeholder="Contraseña"
            type="password"
            value="123456"
          />
        </div>
        <div
          className="form-group"
        >
          <input
            className="btnSubmit"
            type="submit"
            value="Login"
          />
        </div>
      </form>
    </div>
    <div
      className="col-md-6 login-form-2"
    >
      <h3>
        Register
      </h3>
      <form
        onSubmit={[Function]}
      >
        <div
          className="form-group"
        >
          <input
            className="form-control"
            name="rName"
            onChange={[Function]}
            placeholder="Nombre"
            type="text"
            value="Moises"
          />
        </div>
        <div
          className="form-group"
        >
          <input
            className="form-control"
            name="rEmail"
            onChange={[Function]}
            placeholder="Correo"
            type="email"
            value="moisesolivar16@gmail.cl"
          />
        </div>
        <div
          className="form-group"
        >
          <input
            className="form-control"
            name="rPassword"
            onChange={[Function]}
            placeholder="Contraseña"
            type="password"
            value="123456"
          />
        </div>
        <div
          className="form-group"
        >
          <input
            className="form-control"
            name="rPassword2"
            onChange={[Function]}
            placeholder="Repita la contraseña"
            type="password"
            value="123456"
          />
        </div>
        <div
          className="form-group"
        >
          <input
            className="btnSubmit"
            type="submit"
            value="Register"
          />
        </div>
      </form>
    </div>
  </div>
</div>
`;
